Add WORKERFS support

LostBeard opened the pull request:
I'm not sure if this is of interest but I have enabled support for WORKERFS mounting and unmounting.   I am still learning GitHub. If there is an issue with my pull request please let me know. Thank you for the awesome work.

Quick example that uses WORKERFS for the input file.
```js
const transcodePicker = async () => {
    let fileHandle;
    const pickerOpts = {
        types: [
            {
                description: "Videos",
                accept: {
                    "video/*": [".mp4", ".m4v", ".webm", ".avi", ".mkv", ".mov", ".wmv"],
                },
            },
        ],
        excludeAcceptAllOption: true,
        multiple: false,
    };
    [fileHandle] = await window.showOpenFilePicker(pickerOpts);
    let file = await fileHandle.getFile();
    const inputDir = '/input';
    const inputFile = `${inputDir}/${file.name}`;
    await ffmpeg.createDir(inputDir);
    await ffmpeg.mount('WORKERFS', {
        files: [ file ],
    }, inputDir);
    await ffmpeg.exec(['-i', inputFile, 'output.mp4']);
    const output = await ffmpeg.readFile('output.mp4');
    videoEl.src = URL.createObjectURL(new Blob([output.buffer], { type: 'video/mp4' }));
    await ffmpeg.unmount(inputDir);
    await ffmpeg.deleteDir(inputDir);
};
```

LostBeard commented:
Just updated the branch for this pull request to support calling ffmpeg.mount() with the name of the filesystem type to mount so that it can support any additional filesystems that may be enabled for the build. I also updated my usage example in the first post here.

LostBeard commented:
[Simple demo](https://lostbeard.github.io/SpawnDev.BlazorJS.FFmpegWasm/) of a Blazor WASM app using ffmpeg.wasm with WORKERFS via this code.

jeromewu commented:
LGTM, thanks!

ashuvssut commented:
I think WORKERFS uses File System Web API internally. 

if I am not wrong with what I said above, then I think WORKERFS might not work in Firefox or Safari because [File System web API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API) is not fully supported in these two browsers
If that's the case, what's the fallback? do we use RAM for fallback?

LostBeard commented:
@ashuvssut You are wrong. WORKERFS in ffmpeg.wasm works fine in Firefox...unless you are using some intentionally crippled OS by Apple, in which case about 90% of stuff does not work.

WORKERFS is optional and you have to manually use it in your code. Just like any other development, code, test, fix. What is YOUR problem and what does it have to do with THIS pull request for an optional feature?

LostBeard commented:
@ashuvssut [WORKERFS](https://emscripten.org/docs/api_reference/Filesystem-API.html#filesystem-api-workerfs)

ashuvssut commented:
> WORKERFS is optional, 
Oh I didnt know that. Thankyou very much for this implementation. Its this also tested on Safari?

>What is YOUR issue and what does it have to do with THIS pull request for an optional feature?
I want to just process large video files using ffmpeg wasm

LostBeard commented:
> Is this also fully working on Safari?

Safari for what OS? 

Safari on iOS is significantly different than Safari on OSx. I have neither at this time so I cannot answer that. You will have to test on Safari to determine compatibility. 

> I want to just process large video files using ffmpeg wasm

Cool beans. That is why WORKERFS was added. 👍

ashuvssut commented:
> Safari for what OS?

OSx only.

I am trying to test with a 19GB mkv file to convert to mp4 on Safari

It has been around more than 20mins since starting the transcoding process. The Maximum memory consumed seems to be steadily growing starting from 1.30GB to 1.56GB currently after 20mins. Does this mean that the RAM is being used instead of WORKERFS? is RAM actually the fallback for WORKERFS? Total transcoded video time is 214secs

## Here's the last screenshot of the devtools 
20mins taken for processing. RAM consumption steadily increased from  1.30GB to 1.70GB. Total transcoded video duration is 214secs


- <img width="897" alt="image" src="https://github.com/ffmpegwasm/ffmpeg.wasm/assets/60546840/a17f8e89-9245-4f0b-8986-56fb2d99be8c">


https://github.com/ffmpegwasm/ffmpeg.wasm/assets/60546840/22246a21-0ca6-44a0-be04-7f398a605244

LostBeard commented:
When WORKERFS is used, it allows mounting [File](https://developer.mozilla.org/en-US/docs/Web/API/File) or [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob) objects into a virtual folder ffmpeg.wasm can read directly from. Without it, you need to copy all of the input files into the default file system [MEMFS](https://emscripten.org/docs/api_reference/Filesystem-API.html#memfs), an in-memory file system.

WORKERFS files are read only and can only be used for ffmpeg.wasm's input.  

ffmpeg.wasm will still use the default file system (MEMFS) for writing the output. This could be causing the increase in memory over time you are seeing.

I updated my [Blazor WASM ffmpeg.wasm demo](https://lostbeard.github.io/SpawnDev.BlazorJS.FFmpegWasm/). It now uses the current release version and I added a checkbox for toggling WORKERFS on or off.

LostBeard commented:
The processing is all being done in workers. Are both browsers showing memory usage for those also?

KyGuy2002 commented:
Hey, thanks for all this awesome work LostBeard!

I am looking into this, and wondering how this would let you download or view the output file?  If WORKERFS allows mounting files greater than 2gb, but is read only, then wouldnt ffmpeg write the output into MEMFS, which would also be more than 2gb?

Is this limitation for specific files, or total size?

I apologize if I missed something.  Thanks!

LostBeard commented:
Correct. Adding **WORKERFS** allows using very large files as **input** without having to load the entire file into memory, and is read-only. The **output** is still written to MEMFS. Depending on what you are trying to do, output size may not be an issue. If you are just creating a GIF from a small section of a very large video MEMFS is fine. 

Maybe [IDBFS](https://emscripten.org/docs/api_reference/Filesystem-API.html#filesystem-api-idbfs) could be added as an alternative to MEMFS. I have not had a chance to look into it much yet.

LostBeard commented:
I do not know what the exact limitations are in terms of GB and input. I would imagine that also depends on what you are trying to do in terms of processing.

KyGuy2002 commented:
Thanks for the info.  I was able to get WORKERFS working for the input using your example above, but I am trying to concat many clips together so the output file will always be large.  Do you know if it's currently possible at all with WASM to do this?

I was reading a little about streaming/chunk the output from WASM to JS, but not sure the details or if that's even possible?

Thank you!

jj449 commented:
@LostBeard  Thank you so much !

2-fly-4-ai commented:
> Thanks for the info. I was able to get WORKERFS working for the input using your example above, but I am trying to concat many clips together so the output file will always be large. Do you know if it's currently possible at all with WASM to do this?
> 
> I was reading a little about streaming/chunk the output from WASM to JS, but not sure the details or if that's even possible?
> 
> Thank you!

Any update on this?
